### PR TITLE
port physlr split-minimizers to c++

### PR DESCRIPTION
**Pull request by jowong4**
We find that using split minimizers gave better scaffold to molecule mapping than without. In particular, this should resolve one of them problems we saw which was the same scaffold mapping to multiple location.

### CODE REVIEW

File: src/physlr-split-minimizers.cc [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+		printHelpDialog();
+		exit(EXIT_FAILURE);
+	}
+
+	// constuct minimizers to barcodes
+	// barcode to ID table (index in vector)
+	tsl::robin_map<std::string, BarcodeID> barcodes;
+
+	// barcodeID (index) to minimizer vector of vector
+	// Note: Because a vector isn't a set, the input cannot have duplicates.
```
Comment (lcoombe) [inline src/physlr-split-minimizers.cc]: It is possible that the input has duplicate minimizers, right? Would that be an issue if that happened (other than the increased memory)

File: src/physlr-split-minimizers.cc [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+#endif
+	std::cerr << "Memory usage: " << double(memory_usage()) / double(1048576) << "GB" << std::endl;
+}
+
+void
+findMoleculesPerBarcode(bxToMolIdx_t& bxToMolIdx, const graph_t& g)
+{
+	auto vertexItRange = boost::vertices(g);
+	for (auto vertexIt = vertexItRange.first; vertexIt != vertexItRange.second; ++vertexIt) {
+		std::string pattern = R"((\S*?)_\d+_\d+$)";
```
[the remark this answers is not in the document]
Comment (jowong4) [reply src/physlr-split-minimizers.cc]: I just copied and pasted the regex. I could see if that works the same

File: src/physlr-split-minimizers.cc [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+	} else if (version != 0) {
+		printHelpDialog();
+		exit(EXIT_SUCCESS);
+	}
+
+#if _OPENMP
+	omp_set_num_threads(opt::threads);
+#endif
+
+	// Stores fasta input file names
```
[the remark this answers is not in the document]
Comment (jowong4) [reply src/physlr-split-minimizers.cc]: was copying from another physlr file

File: src/physlr-split-minimizers.cc (hunk 1 above)
Comment (jowong4) [reply src/physlr-split-minimizers.cc]: How so. The results should be filtered and merged at that point so there would be no duplicate minimizers per barcode. This is the same assumption that `physlr-overlap` makes.

File: src/physlr-split-minimizers.cc [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+		auto& bxIdx = barcodes[bx];
+		auto& minimizerSet = barcodeToMinimizer[bxIdx];
+
+		for (auto& mol : (*it).second) {
+			// Get Union of minimizers of neighbours
+			auto neighbours = boost::adjacent_vertices(mol, g);
+			std::unordered_set<Minimizer> neighbourMxsUnion;
+			for (auto neighbourItr = neighbours.first; neighbourItr != neighbours.second;
+			     ++neighbourItr) {
+				std::string pattern = R"((\S*?)_(\d+)_(\d+)$)";
```
[the remark this answers is not in the document]
Comment (jowong4) [reply src/physlr-split-minimizers.cc]: Forgot to remove the capture in this one.

File: src/physlr-split-minimizers.cc (hunk 1 above)
Comment (lcoombe) [reply src/physlr-split-minimizers.cc]: Ah! Yup you're right -- I was thinking about the indexlr file with 1 read per line.

File: src/physlr-split-minimizers.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (lcoombe) [reply src/physlr-split-minimizers.cc]: I think (?) it would, but logically having both `*` and `?` is unnecessary

**Comment by jowong4** [comment]
| | Time (hh:mm:ss) | Ram (GB) |
|--|--|--|
|Python|7:25:01|345|
|C++|0:41:02|30|

Based on usr/bin/time

**Comment by lcoombe** [comment]
> Time (hh:mm:ss)	Ram (GB)
> Python 16 threads	7:25:01	345
> C++ 48 threads	0:41:02	30
> Based on usr/bin/time

That's a great improvement! :+1:

File: src/physlr-split-minimizers.cc [hunk 5]
```diff
@@ -399,7 +399,13 @@ main(int argc, char* argv[])
 	sTime = omp_get_wtime();
 #endif
 
+	std::cerr << "Splitting minimizers" << std::endl;
 	bxToMolIdx_t bxToMolIdx;
 	findMoleculesPerBarcode(bxToMolIdx, g);
 	splitMinimizers(bxToMolIdx, barcodeToMinimizer, g, barcodes);
+	std::cerr << "Splitted minimizers" << std::endl;
```
Comment (lcoombe) [inline src/physlr-split-minimizers.cc]:
```suggestion
	std::cerr << "Finished splitting minimizers" << std::endl;
```

**Pull request merged**